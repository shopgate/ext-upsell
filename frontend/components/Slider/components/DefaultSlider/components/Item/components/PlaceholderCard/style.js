import { css } from 'glamor';
import isiOSTheme from '../../../../../../../../helpers/isiOSTheme';

/**
 * Glow animation.
 */
const placeholderCycle = css.keyframes('placeholder-cycle', {
  '0%': {
    backgroundPosition: '-240% 0',
  },
  '100%': {
    backgroundPosition: '240% 0',
  },
});

/**
 * Main wrapper.
 */
const wrapper = css({
  height: 'auto',
  position: 'relative',
  // Glow animation
  ':after': {
    content: '""',
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    background: 'linear-gradient(to right, rgba(255,255,255,0) 0%,rgba(255,255,255,0.35) 29%,rgba(255,255,255,0.35) 30%,rgba(255,255,255,0.35) 48%,rgba(255,255,255,0.35) 52%,rgba(255,255,255,0.35) 70%,rgba(255,255,255,0) 100%)',
    backgroundSize: '60% 100%',
    backgroundRepeat: 'no-repeat',
    animation: `${placeholderCycle} 1s infinite`,
    transform: 'translate3d(0,0,0)',
  },
}).toString();

/**
 * Image placeholder.
 */
const image = css({
  display: 'block',
  width: '100%',
  background: 'rgba(197,197,197,0.2)',
}).toString();

/**
 * Template rows helper.
 * @param {number} titleRows Title rows.
 * @param {boolean} hideName Hide name.
 * @returns {string}
 */
const getParagraphRemplateRows = (titleRows, hideName) => {
  if (hideName) {
    return '0';
  }

  return `repeat(${titleRows}, 1fr)`;
};

/**
 * Paragraph margin bottom helper.
 * @param {boolean} hideName Hide name.
 * @param {boolean} hidePrice Hide price.
 * @returns {string}
 */
const getParagraphMarginBottom = (hideName, hidePrice) => {
  const isIOS = isiOSTheme();
  if (hideName && hidePrice) {
    return '0';
  }

  if (hideName) {
    return isIOS ? '1.8em' : 'calc(1.75em - 1px)';
  }

  if (hidePrice) {
    return '0';
  }

  return isIOS ? '1.8em' : '1.75em';
};

/**
 * Paragraph css factory.
 * @param {number} titleRows Expected title rows.
 * @param {boolean} hideName Hide name.
 * @param {boolean} hidePrice Hide price.
 * @returns {string}
 */
const paragraph = (titleRows, hideName, hidePrice) => css({
  display: 'grid',
  background: 'linear-gradient(to bottom, rgba(197,197,197,0) 0%,rgba(197,197,197,0) 20%,rgba(197,197,197,0.4) 20.1%,rgba(197,197,197,0.4) 50%,rgba(197,197,197,0.4) 80%,rgba(197,197,197,0) 80.1%,rgba(197,197,197,0) 100%) repeat',
  backgroundSize: `100% ${100 / titleRows}%`,
  gridTemplateRows: getParagraphRemplateRows(titleRows, hideName),
  marginBottom: getParagraphMarginBottom(hideName, hidePrice), // Price line height
  color: 'transparent',
  fontSize: isiOSTheme() ? 14 : 'inherit',
}).toString();

export default {
  wrapper,
  image,
  paragraph,
};
