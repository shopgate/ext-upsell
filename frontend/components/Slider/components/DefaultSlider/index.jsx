import React from 'react';
import PropTypes from 'prop-types';
import Slider from '@shopgate/pwa-common/components/Slider';
import Item from './components/Item';
import getStyles from '../../../../styles/slider';

const styles = getStyles();

/**
 * Renders default Slider.
 * @param {Object} props Props.
 * @param {Array} props.products Products.
 * @param {boolean} showPrice Show price.
 * @param {boolean} showName Show name.
 * @returns {JSX}
 */
const DefaultSlider = ({ products, showPrice, showName }) => (
  <Slider
    slidesPerView={2.3}
    classNames={{
      container: styles.defaultSliderContainer,
    }}
  >
    {
      products.map(product => (
        <Item
          product={product}
          key={product.id}
          showPrice={showPrice}
          showName={showName}
        />
      ))
    }
  </Slider>
);

DefaultSlider.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape({})),
  showName: PropTypes.bool,
  showPrice: PropTypes.bool,
};

DefaultSlider.defaultProps = {
  products: [],
  showPrice: false,
  showName: false,
};

export default DefaultSlider;
