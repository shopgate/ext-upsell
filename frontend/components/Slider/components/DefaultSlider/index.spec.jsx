import React, { Component as MockedComponent } from 'react';
import { mount } from 'enzyme';
import DefaultSlider from './index';
import IOSProductCard from './components/Item/components/ios11/ProductCard';
import GMDProductCard from './components/Item/components/gmd/ProductCard';
import PlaceholderCard from './components/Item/components/PlaceholderCard';

jest.mock('@shopgate/pwa-common/components/Slider', () => class extends MockedComponent {
  // eslint-disable-next-line require-jsdoc
  render() {
    return <div>{ this.props.children }</div>;
  }
  // eslint-disable-next-line require-jsdoc
  static Item({ children }) {
    return (<div>{ children }</div>);
  }
});

let mockedIsiOSTheme = false;
jest.mock('../../../../helpers/isiOSTheme', () => () => mockedIsiOSTheme);

describe('DefaultSlider', () => {
  it('should render gmd slider', () => {
    const products = {
      mockedId: {
        id: 'mockedId',
        name: 'Mocked Product',
        price: {
          unitPrice: 100,
          currency: 'EUR',
        },
      },
    };
    const component = mount(<DefaultSlider products={products} productIds={['mockedId']} />);
    expect(component.find('Item').exists()).toBe(true);
    expect(component.find(GMDProductCard).exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });

  it('should render ios slider', () => {
    mockedIsiOSTheme = true;
    const products = {
      mockedId: {
        id: 'mockedId',
        name: 'Mocked Product',
        price: {
          unitPrice: 100,
          currency: 'EUR',
        },
      },
    };
    const component = mount(<DefaultSlider products={products} productIds={['mockedId']} />);
    expect(component.find('Item').exists()).toBe(true);
    expect(component.find(IOSProductCard).exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });

  it('should render a placeholder', () => {
    const products = {};
    const component = mount(<DefaultSlider products={products} productIds={['mockedId']} />);
    expect(component.find('Item').exists()).toBe(true);
    expect(component.find(PlaceholderCard).exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});
