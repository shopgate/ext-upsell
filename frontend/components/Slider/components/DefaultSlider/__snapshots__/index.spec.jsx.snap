// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultSlider should render a placeholder 1`] = `
<DefaultSlider
  productIds={
    Array [
      "mockedId",
    ]
  }
  products={Object {}}
  showName={false}
  showPrice={false}
>
  <_class
    classNames={
      Object {
        "container": "css-13azwyo",
      }
    }
    slidesPerView={2.3}
  >
    <div>
      <Item
        key="placeholder-mockedId"
        product={null}
        showName={false}
        showPrice={false}
      >
        <Item>
          <div>
            <Card
              className="css-d1ao1r"
            >
              <div
                className="css-1l0if7f css-d1ao1r"
              >
                <PlaceholderCard
                  hideName={true}
                  hidePrice={true}
                  hideRating={true}
                  product={null}
                  titleRows={2}
                >
                  <div
                    className="css-1dwnok8"
                  >
                    <div
                      data-test-id="upselling-placeholder-image"
                    >
                      <img
                        alt=""
                        className="css-1kzppb9"
                        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAbgAAAG4CAQAAACdwzo4AAADg0lEQVR42u3TMQ0AAAzDsJU/6f0F0MuGECk5YCYSgOHAcIDhwHCA4cBwYDjAcGA4wHBgODAcYDgwHGA4MBxgODAcGA4wHBgOMBwYDgwHGA4MBxgODAcYDgwHhgMMB4YDDAeGA8MBhgPDAYYDwwGGA8OB4QDDgeEAw4HhwHCA4cBwgOHAcGA4wHBgOMBwYDjAcGA4MBxgODAcYDgwHBgOMBwYDjAcGA4wHBgODAcYDgwHGA4MB4YDDAeGAwwHhgMMB4YDwwGGA8MBhgPDgeEAw4HhAMOB4cBwgOHAcIDhwHCA4cBwYDjAcGA4wHBgODAcYDgwHGA4MBxgODAcGA4wHBgOMBwYDgwHGA4MBxgODAcYDgwHhgMMB4YDDAeGA8MBhgPDAYYDw4HhAMOB4QDDgeEAw4HhwHCA4cBwgOHAcGA4wHBgOMBwYDjAcGA4MBxgODAcYDgwHBgOMBwYDjAcGA4wHBgODAcYDgwHGA4MB4YDDAeGAwwHhgPDAYYDwwGGA8MBhgPDgeEAw4HhAMOB4cBwgOHAcIDhwHCA4cBwYDjAcGA4wHBgODAcYDgwHGA4MBxgODAcGA4wHBgOMBwYDgwHGA4MBxgODAeGAwwHhgMMB4YDDAeGA8MBhgPDAYYDw4HhAMOB4QDDgeEAw4HhwHCA4cBwgOHAcGA4wHBgOMBwYDjAcGA4MBxgODAcYDgwHBgOMBwYDjAcGA4MJwEYDgwHGA4MBxgODAeGAwwHhgMMB4YDwwGGA8MBhgPDAYYDw4HhAMOB4QDDgeHAcIDhwHCA4cBwgOHAcGA4wHBgOMBwYDgwHGA4MBxgODAcYDgwHBgOMBwYDjAcGA4MBxgODAcYDgwHhgMMB4YDDAeGAwwHhgPDAYYDwwGGA8OB4QDDgeEAw4HhAMOB4cBwgOHAcIDhwHBgOMBwYDjAcGA4wHBgODAcYDgwHGA4MBwYDjAcGA4wHBgODAcYDgwHGA4MBxgODAeGAwwHhgMMB4YDwwGGA8MBhgPDAYYDw4HhAMOB4QDDgeHAcIDhwHCA4cBwgOHAcGA4wHBgOMBwYDgwHGA4MBxgODAcGA4wHBgOMBwYDjAcGA4MBxgODAcYDgwHhgMMB4YDDAeGAwwHhgPDAYYDwwGGA8OB4QDDgeEAw4HhAMOB4cBwgOHAcIDhwHBgOMBwYDjAcGA4MBxgODAcYDgwHNAedDYBuQI5h3IAAAAASUVORK5CYII="
                      />
                    </div>
                  </div>
                </PlaceholderCard>
              </div>
            </Card>
          </div>
        </Item>
      </Item>
    </div>
  </_class>
</DefaultSlider>
`;

exports[`DefaultSlider should render gmd slider 1`] = `
<DefaultSlider
  productIds={
    Array [
      "mockedId",
    ]
  }
  products={
    Object {
      "mockedId": Object {
        "id": "mockedId",
        "name": "Mocked Product",
        "price": Object {
          "currency": "EUR",
          "unitPrice": 100,
        },
      },
    }
  }
  showName={false}
  showPrice={false}
>
  <_class
    classNames={
      Object {
        "container": "css-13azwyo",
      }
    }
    slidesPerView={2.3}
  >
    <div>
      <Item
        key="product-mockedId"
        product={
          Object {
            "id": "mockedId",
            "name": "Mocked Product",
            "price": Object {
              "currency": "EUR",
              "unitPrice": 100,
            },
          }
        }
        showName={false}
        showPrice={false}
      >
        <Item>
          <div>
            <Card
              className="css-d1ao1r"
            >
              <div
                className="css-1l0if7f css-d1ao1r"
              >
                <ProductCard
                  hideName={true}
                  hidePrice={true}
                  hideRating={true}
                  product={
                    Object {
                      "id": "mockedId",
                      "name": "Mocked Product",
                      "price": Object {
                        "currency": "EUR",
                        "unitPrice": 100,
                      },
                    }
                  }
                  titleRows={2}
                >
                  <Link
                    className=""
                    disabled={false}
                    href="/trampoline/item/6d6f636b65644964"
                    itemProp="item"
                    itemScope={true}
                    itemType="http://schema.org/Product"
                    tagName="a"
                  >
                    <div
                      aria-hidden={true}
                      className="css-c6ki83 "
                      data-test-id="link: /trampoline/item/6d6f636b65644964"
                      onClick={[Function]}
                      role="link"
                    >
                      <ProductImage
                        alt="Mocked Product"
                        animating={true}
                        forcePlaceholder={false}
                        highestResolutionLoaded={[Function]}
                        itemProp="image"
                        ratio={null}
                        resolutions={
                          Array [
                            Object {
                              "blur": 2,
                              "height": 50,
                              "width": 50,
                            },
                            Object {
                              "height": 440,
                              "width": 440,
                            },
                          ]
                        }
                        src={null}
                      >
                        <div
                          className="css-y5kkpq"
                        >
                          <div
                            className="css-15j45j0"
                            data-test-id="placeHolder"
                          >
                            <Placeholder
                              className="css-19rc2oq"
                            >
                              <Icon
                                className="css-19rc2oq"
                                color={null}
                                content="<circle cx=\\"12\\" cy=\\"12\\" r=\\"3.2\\"/><path d=\\"M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z\\"/><path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"/>"
                                size="inherit"
                                viewBox="0 0 24 24"
                              >
                                <svg
                                  className="css-yys9hb css-19rc2oq"
                                  dangerouslySetInnerHTML={
                                    Object {
                                      "__html": "<circle cx=\\"12\\" cy=\\"12\\" r=\\"3.2\\"/><path d=\\"M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z\\"/><path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"/>",
                                    }
                                  }
                                  style={
                                    Object {
                                      "fill": null,
                                      "fontSize": "inherit",
                                    }
                                  }
                                  viewBox="0 0 24 24"
                                  xmlns="http://www.w3.org/2000/svg"
                                />
                              </Icon>
                            </Placeholder>
                          </div>
                        </div>
                      </ProductImage>
                    </div>
                  </Link>
                </ProductCard>
              </div>
            </Card>
          </div>
        </Item>
      </Item>
    </div>
  </_class>
</DefaultSlider>
`;

exports[`DefaultSlider should render ios slider 1`] = `
<DefaultSlider
  productIds={
    Array [
      "mockedId",
    ]
  }
  products={
    Object {
      "mockedId": Object {
        "id": "mockedId",
        "name": "Mocked Product",
        "price": Object {
          "currency": "EUR",
          "unitPrice": 100,
        },
      },
    }
  }
  showName={false}
  showPrice={false}
>
  <_class
    classNames={
      Object {
        "container": "css-13azwyo",
      }
    }
    slidesPerView={2.3}
  >
    <div>
      <Item
        key="product-mockedId"
        product={
          Object {
            "id": "mockedId",
            "name": "Mocked Product",
            "price": Object {
              "currency": "EUR",
              "unitPrice": 100,
            },
          }
        }
        showName={false}
        showPrice={false}
      >
        <Item>
          <div>
            <Card
              className="css-d1ao1r"
            >
              <div
                className="css-1l0if7f css-d1ao1r"
              >
                <ProductCard
                  hideName={true}
                  hidePrice={true}
                  hideRating={true}
                  product={
                    Object {
                      "id": "mockedId",
                      "name": "Mocked Product",
                      "price": Object {
                        "currency": "EUR",
                        "unitPrice": 100,
                      },
                    }
                  }
                  titleRows={2}
                >
                  <Link
                    className="css-b4pchx"
                    disabled={false}
                    href="/trampoline/item/6d6f636b65644964"
                    itemProp="item"
                    itemScope={true}
                    itemType="http://schema.org/Product"
                    tagName="a"
                  >
                    <div
                      aria-hidden={true}
                      className="css-c6ki83 css-b4pchx"
                      data-test-id="link: /trampoline/item/6d6f636b65644964"
                      onClick={[Function]}
                      role="link"
                    >
                      <ProductImage
                        alt="Mocked Product"
                        animating={true}
                        classNames={
                          Object {
                            "container": null,
                            "glowContainer": null,
                            "imageContainer": null,
                          }
                        }
                        forcePlaceholder={false}
                        highestResolutionLoaded={[Function]}
                        itemProp="image"
                        ratio={null}
                        resolutions={
                          Array [
                            Object {
                              "blur": 2,
                              "height": 50,
                              "width": 50,
                            },
                            Object {
                              "height": 440,
                              "width": 440,
                            },
                          ]
                        }
                        src={null}
                      >
                        <div
                          className="css-3hjqnz"
                        >
                          <div
                            className="css-y5kkpq"
                          >
                            <div
                              className="css-15j45j0"
                              data-test-id="placeHolder"
                            >
                              <Placeholder
                                className="css-19rc2oq"
                              >
                                <Icon
                                  className="css-19rc2oq"
                                  color={null}
                                  content="<circle cx=\\"12\\" cy=\\"12\\" r=\\"3.2\\"/><path d=\\"M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z\\"/><path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"/>"
                                  size="inherit"
                                  viewBox="0 0 24 24"
                                >
                                  <svg
                                    className="css-yys9hb css-19rc2oq"
                                    dangerouslySetInnerHTML={
                                      Object {
                                        "__html": "<circle cx=\\"12\\" cy=\\"12\\" r=\\"3.2\\"/><path d=\\"M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z\\"/><path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"/>",
                                      }
                                    }
                                    style={
                                      Object {
                                        "fill": null,
                                        "fontSize": "inherit",
                                      }
                                    }
                                    viewBox="0 0 24 24"
                                    xmlns="http://www.w3.org/2000/svg"
                                  />
                                </Icon>
                              </Placeholder>
                            </div>
                          </div>
                        </div>
                      </ProductImage>
                    </div>
                  </Link>
                </ProductCard>
              </div>
            </Card>
          </div>
        </Item>
      </Item>
    </div>
  </_class>
</DefaultSlider>
`;
